bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/DependencyTracker.java

```diff
@@ -31,40 +29,48 @@
  *
  * @author Jonathan Gilbert - Initial contribution
  */
-@Component(immediate = true, service = DependencyTracker.class)
```

Comment:
Can you help me with this? Will it be necessary for every script automation bundle to implement it's own `DependencyTracker` service after this change?

Reply:
It won't be necessary (it's entirely optional), but it will be required _if_ they want explicit dependency tracking - e.g. script reloading if dependencies change (no addon actually does track dependencies except the JS one though afaik). You can see how the `DependencyTracker` is actually added in the associated JS PR.

- https://github.com/openhab/openhab-addons/pull/11719

Reply:
Thanks for the link. I saw your changes over there. That's why I am asking.

Please correct me if I am wrong. For now the dependency tracking is available by default for everyone automation script bundle. With this PR it will be turned of which will be a great loss if we do not implement it immediately for every existing extension.

Or does it work only for JS-based automation?

Reply:
Currently it only works for JS-based automation.

The reason for this is that there are two pieces to be implemented for it to work:
- Monitoring of any library file changes (where this class is involved)
- Establishing the dependency graph of which script depends on which library (so we can determine which scripts to reload if a library file changes)

Whilst _technically_ all languages currently get the first part for free (because currently all libraries are in the same source root), only the JS addon actually implements the second part. This second part can be implemented by any language by notifying the dependency tracker of explicit dependencies, which the JS addon does by hooking into the module loading system and recording them as they happen.

So there is no loss of existing functionality here; only slightly more work (at the benefit of much greater flexibility) for other languages to add dependency tracking at some point.